xayn-ai/src/coi/key_phrase.rs

```diff
@@ -137,24 +139,27 @@ impl RelevanceMap {
 }
 
 /// Unifies the key phrases and candidates.
-fn unify<F>(
+fn unify(
     mut key_phrases: BTreeSet<KeyPhrase>,
     candidates: &[String],
-    smbert: F,
-) -> BTreeSet<KeyPhrase>
-where
-    F: Fn(&str) -> Result<Embedding, Error>,
-{
-    for candidate in candidates {
-        if !key_phrases.contains(candidate) {
-            if let Ok(Ok(candidate)) =
-                smbert(candidate).map(|point| KeyPhrase::new(candidate, point))
-            {
-                key_phrases.insert(candidate);
-            }
-        }
-    }
+    smbert: impl Fn(&str) -> Result<Embedding, Error> + Sync,
+) -> BTreeSet<KeyPhrase> {
+    #[cfg(not(feature = "multithreaded"))]
+    let candidates = candidates.iter();
+    #[cfg(feature = "multithreaded")]
+    let candidates = candidates.into_par_iter();
+
+    let mut candidates = candidates
+        .filter(|candidate| !key_phrases.contains(*candidate))
+        .map(clean_key_phrase)
+        .filter_map(|candidate| {
+            smbert(&candidate)
+                .and_then(|point| KeyPhrase::new(candidate, point).map_err(|e| e.into()))
+                .ok()
```

Comment:
The `map_err` -> `ok` is a bit strange, I mean the only reason why it happens to work is because there happens to be an error conversion for reasons unrelated to this code.

Maybe:

```suggestion
            smbert(&candidate).ok()
                .and_then(|point| KeyPhrase::new(candidate, point).ok())
```
or
```suggestion
            let point = smbert(&candidate).ok()?;
            KeyPhrase::new(candidate, point).ok()
```

would be slightly better.

Reply:
I don't understand the point on `map_err`, is not only working by a random chance, the signature request the same type so we do the conversion. Having an `ok` inside the `and_then` was more weird to me and I'll prefer to leave the decision on what to do to the outer code.
I'll go for the last option it is indeed better.

Reply:
Because in other place (discovery engine) we are logging them. I'll prefer to keep the base method returning an error as it should and then have the "higher" level function decide how to deal with it.